### DISCUSSION TITLE
R: Autocompletion inconsistency with RStudio

### DISCUSSION THREAD

**Discussion by FinYang** [posit-dev/positron#4442] [Feedback]
I recently migrated from RStudio and am a bit uncomfortable with how Positron's autocompletion behaves. I am not sure if there is really something going on or if it's really just my own habits (so listing them here in a discussion instead of issues).

## 1. Backing out of suggestion widgets with the left arrow key

In RStudio, when I finish typing an object and don't need the suggestion anymore, I usually like to move with my (left) arrow key to make the suggestion box go away, so I can then move to other lines (up and down arrow key).

![Kapture 2024-08-22 at 23 35 20](https://github.com/user-attachments/assets/a34945ab-b818-4451-8555-c190d88db1ae)

In Positron, the left and right arrow keys are "kidnapped" by the suggestion box, so I have to struggle to get out of it

![Kapture 2024-08-22 at 23 37 03](https://github.com/user-attachments/assets/7a47fe08-80b0-455f-a6cb-8ec47fda4268)

Maybe this is my own habit but using left arrow key to show alternative suggestions with fewer matching characters might not be of much use, because the same function is more naturally supported by BackSpace.

I would like to have an option such that left and right arrow keys close suggestions, but up and down arrow keys can still be used to go between suggestions, just like how RStudio did it.

## 2. Matching the beginning of the suggestions

In RStudio, only suggestions that matches from the beginning are included,

![Kapture 2024-08-22 at 23 50 50](https://github.com/user-attachments/assets/141a1aac-3064-4984-a5ab-64079c41de67)

but in Positron, the only option is to "Match On Word Start", so if the suggestion has multiple words separated by dots or with camel case, suggestions matched from the middle will also be included:

![Kapture 2024-08-22 at 23 55 13](https://github.com/user-attachments/assets/162e5ee0-a07a-433b-9b91-06df62f05efc)

That's just too much for me. It would be great if there is an option to disable this.

## 3. Closing suggestions when functions are typed out in full

In RStudio, suggestions automatically close when the functions are typed out in full/when there is only one suggestion that matches exactly (reusing the gif).

![Kapture 2024-08-22 at 23 50 50](https://github.com/user-attachments/assets/141a1aac-3064-4984-a5ab-64079c41de67)

In Positron, this is not the case

![Kapture 2024-08-23 at 00 00 27](https://github.com/user-attachments/assets/4648c2b7-a04c-4dbf-8194-60551d78234e)

This is an issue when I just want to type out the name of the function in the console, hit enter and see the source code. In Positron, I would have to get out of the suggestion box somehow, with some effort because of the first issue mentioned above,
or hit enter and delete the parentheses myself, somewhat similar to what is described in #1563.

If I want to view the source code, the second issue mentioned above can also be a problem, even if issue no.3 is addressed because there are just too many suggestions and the suggestion box wouldn't close by itself. This also relates to the last issue I have encountered so far:

## 4. Matching suggestions without gaps

In RStudio:

![Kapture 2024-08-23 at 00 07 34](https://github.com/user-attachments/assets/f0453c11-3a9d-46f7-b50d-73fdfb32c8e4)

In Positron:

![Kapture 2024-08-23 at 00 08 17](https://github.com/user-attachments/assets/326c8f19-0e5c-4695-89ab-53e4d5637a4c)

Any suggestions matching partially the pattern are provided. This is especially an issue for short object names because obviously in the example above I would want to type `rnorm` and not `randomForest`. I would want to have the option for the autocompletion to be doing more "autocompleting" than "searching", if that makes any sense.

## So...

I guess I pretty much want the autocompletion in Positron to behave exactly like RStudio (or have the option of doing so), if that's not too greedy...

**Comment by FinYang** [comment]
To fix issue no. 1, add the following keybindings to keyboard shortcuts (Command Palatte -> Preferences: Open Keyboard Shortcuts (JSON))

``` json
    {
        "command": "runCommands",
        "key": "left",
        "args": {
            "commands": [
                "hideSuggestWidget",
                "cursorLeft"
            ]
        },
        "when": "suggestWidgetVisible && textInputFocus"
    },
    {
        "command": "runCommands",
        "key": "right",
        "args": {
            "commands": [
                "hideSuggestWidget",
                "cursorRight"
            ]
        },
        "when": "suggestWidgetVisible && textInputFocus"
    }
```

**Comment by juliasilge** [comment]
Thanks for this detailed feedback! We do have [quite a lot of work tracked for improving completions for R](https://github.com/posit-dev/positron/issues?q=is:issue+is:open+label:%22lang:+r%22+label:%22area:+completions%22) but I've opened these issues to address some of the specifics you raised:


- https://github.com/posit-dev/positron/issues/4488
- https://github.com/posit-dev/positron/issues/4489
- https://github.com/posit-dev/positron/issues/4490
